### PR TITLE
Implement tgz#save

### PR DESCRIPTION
**Pull request by Sammers21**
_Closes #21

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#25](https://github.com/artipie/helm-adapter/issues/25) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/730297f4-9916-41af-b9ae-f8c80dc63211); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
This pull request [#25](https://github.com/artipie/helm-adapter/issues/25) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CT2E6TK9B/6056ea37-54c7-4b88-b6b8-c7ce4fac76d5); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
@0crat[/z](https://www.0crat.com/u/0crat) I see that you unassigned this issue; the order is still assigned to @victornoel[/z](https://www.0crat.com/u/victornoel) though; to cancel the order use `refuse`, as in [§6](http://www.zerocracy.com/policy.html#6)

File: .pdd [hunk 1]
```diff
@@ -3,6 +3,6 @@
 --exclude target/**/*
 --exclude src/test/resources/**/*
 --exclude .idea/**/*
---rule min-words:10
```
Comment (victornoel) [inline .pdd:6]: @g4s8 consider this change when accepting or refusing the PR

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 2]
```diff
@@ -113,9 +119,24 @@ public String file(final String name) {
...
     public Single<Key> save(final Storage storage) {
-        // @todo #12:30min Save the archive into Asto.
-        //  For now this method is not implemented. The archive should be saved with a key name,
-        //  obtained from TgzArchive#name().
-        return Single.error(new IllegalStateException("Not Implemented"));
+        final int eightkb = 8 * 1024;
+        final int resid = this.content.length % eightkb;
+        final int last = resid == 0 ? 0 : 1;
+        final int chunks = this.content.length / eightkb + last;
+        final ArrayList<ByteBuffer> arr = new ArrayList<>(chunks);
```
Comment (victornoel) [inline src/main/java/com/artipie/helm/TgzArchive.java:126]: @Sammers21 why is the content split in chunks and so on here and not just passed to `Content.From`?

If this is really needed, please isolate this behaviour in a separate class that either:
- implements `Content`, for example `new Content.Chunked(bytes)`.
- implements `Flowable`, for example `new Content.From(new Chunked(bytes))`

File: .pdd (hunk 1 above)
Comment (victornoel) [inline .pdd:6]: @Sammers21 please remove this as it is unrelated to this PR

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:126]: I would never reuse it and it is already isolated

File: .pdd (hunk 1 above)
Comment (victornoel) [reply .pdd]: @Sammers21 it doesn't change the fact it has nothing to do with this PR. Also, by the way, the commit messages are unreadable, they mention issues that are not related to this PR either, so it's usable for me as a REV.

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (victornoel) [reply src/main/java/com/artipie/helm/TgzArchive.java:126]: @g4s8 please decide what to do here

File: .pdd (hunk 1 above)
Comment (victornoel) [reply .pdd]: @Sammers21 for example why is #19 cited in the commits while this is related to #21?

File: src/main/java/com/artipie/helm/TgzArchive.java (hunk 2 above)
Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:126]: @victornoel, are you familiar with [KISS ](https://en.wikipedia.org/wiki/KISS_principle)principle?

Comment (victornoel) [reply src/main/java/com/artipie/helm/TgzArchive.java:126]: @Sammers21 you know, no need to get aggressive or condescending, I'm doing the job I'm being paid for, I have nothing against you.

I know KISS, but I was more concerned about SRP in that case. You have many options in front of you to take my request into account:
- add a todo
- create an issue in asto referencing this PR for extracting this common code in ast and you will get paid for it
- or do it here

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java:126]: @victornoel, when we use this method we don't care about the name, when we use it as `Content` we should make one more `name()` call

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 3]
```diff
@@ -113,9 +121,34 @@ public String file(final String name) {
      * @return Asto location, where archive is save.
      */
     public Single<Key> save(final Storage storage) {
-        // @todo #12:30min Save the archive into Asto.
-        //  For now this method is not implemented. The archive should be saved with a key name,
-        //  obtained from TgzArchive#name().
-        return Single.error(new IllegalStateException("Not Implemented"));
+        final Key.From key = new Key.From(this.name());
+        return new RxStorageWrapper(storage)
+            .save(key, new Content.From(this))
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/TgzArchive.java:126]: @Sammers21 why do we need this `new Content.From(this)`? `this` already implements `Content` interfaces, the wrapper doesn't do anything useful:
```java
        public From(final long size, final Publisher<ByteBuffer> publisher) {
            this(Optional.of(size), publisher);
        }
```

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 4]
```diff
@@ -113,9 +121,34 @@ public String file(final String name) {
...
+    }
+
+    @Override
+    public Optional<Long> size() {
+        return Optional.of(this.content.length).map(Integer::longValue);
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super ByteBuffer> subscriber) {
+        final int eightkb = 8 * 1024;
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/TgzArchive.java:137]: @Sammers21 this can be moved to private static final field

File: src/main/java/com/artipie/helm/TgzArchive.java [hunk 5]
```diff
@@ -113,9 +121,34 @@ public String file(final String name) {
...
+        final int chunks = this.content.length / eightkb + last;
+        final ArrayList<ByteBuffer> arr = new ArrayList<>(chunks);
+        for (int idx = 0; idx < chunks; idx += 1) {
+            final byte[] bytes;
+            if (idx == chunks - 1 && last == 1) {
+                bytes = new byte[resid];
+            } else {
+                bytes = new byte[eightkb];
+            }
+            System.arraycopy(this.content, idx * eightkb, bytes, 0, bytes.length);
```
Comment (g4s8) [inline src/main/java/com/artipie/helm/TgzArchive.java:149]: @Sammers21 you actually can avoid creating new arrays for chunks by wrapping origin `byte[]` array to `ByteBuffer` and creating chanked buffers using `slice()` method of `ByteBuffer` - it'll be much faster because it doesn't copy anything, rather creates buffers for same byte array reference, and it'll consume twice less memory.

Comment (Sammers21) [reply src/main/java/com/artipie/helm/TgzArchive.java]: @g4s8, you actually can't use `slice()`, since both versions of slice are not JDK 8 compatible

**Comment by Sammers21** [comment]
@g4s8 , review the PR one more time please

**Comment by 0crat** [comment]
@sereshqua[/z](https://www.0crat.com/u/sereshqua) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by sereshqua** [comment]
@0crat quality good
